meds_interp/knn_raw_code.py

```diff
@@ -0,0 +1,429 @@
+            idx = self.kneighbors(X, self.n_neighbors, return_distance=False)
+            return self.get_uniform_proba(idx)
+        else:  # weights neighbor's labels by inverse distance to the neighbor
+            assert self.weights == "distance"
+            neigh_dist, neigh_ind = self.kneighbors(X, self.n_neighbors, return_distance=True)
+            return self.get_distance_proba(neigh_dist, neigh_ind)
+
+    def concat_data(self, X: list[np.array] | pl.DataFrame):
+        embeddings = []
+        if isinstance(X, pl.DataFrame):
+            for modality in self.modalities:
+                embed_array = np.asarray(X[modality].to_list())
+                embeddings.append(embed_array)
+        else:
+            embeddings = X
+        return np.concatenate(embeddings, axis=1)
+
+    def fit_preprocess(self, X: pl.DataFrame):
+        if self.preprocess == Preprocess_Type.NORM_AFTER_CONCAT:
+            self.scaler = Normalizer()
+            self.scaler.fit(X)
+        elif self.preprocess == Preprocess_Type.NORM_SEPERATLY:
+            self.scalers = []
+            assert X.shape[1] == self.d * 2
+            assert self.modalities.length() == self.modality_weights.length()
+            for i in range(self.modalities.length()):
+                embed_array = np.asarray(X[self.modalities[i]].to_list())
+                scaler = Normalizer()
+                scaler.fit(embed_array)
+                self.scalers.append(scaler * self.modality_weights[i])
+            # for modality in self.modalities:
+            #     embed_array = np.asarray(X[modality].to_list())
+            #     scaler = Normalizer()
+            #     scaler.fit(embed_array)
+            #     self.scalers.append(scaler)
+        else:
+            assert self.preprocess == Preprocess_Type.NONE
+
+    def transform_preprocess(self, X):
+        if self.preprocess == Preprocess_Type.NORM_AFTER_CONCAT:
+            self.scaler = Normalizer()
+            X = self.scaler.transform(X)
+            return X
+        elif self.preprocess == Preprocess_Type.NORM_SEPERATLY:
+            self.scalers = []
+            assert X.shape[1] == self.d * 2
+            assert self.modalities.length() == self.modality_weights.length()
+            for i in range(self.modalities.length()):
+                embed_array = np.asarray(X[self.modalities[i]].to_list())
+                scaler = Normalizer()
+                scaler.transform(embed_array)
+                self.scalers.append(scaler * self.modality_weights[i])
+            # for modality in self.modalities:
+            #     embed_array = np.asarray(X[modality].to_list())
+            #     # scaler = self.scalers[0]
+            #     scaler = Normalizer()
+            #     scaler.transform(embed_array)
+            #     self.scalers.append(scaler)
+            return self.scalers
+        else:
+            assert self.preprocess == Preprocess_Type.NONE
+            return X
```

Comment:
Return X at the end once, and apply the modality weight multiplication right before returning it.

Reply:
```python
modality_lengths = np.array([1,2,3])
modality_weights = np.array([1,.5, .2])
weights = np.repeat(modality_weights, modality_lengths)[None, :]
X * reweight_vector
```

Reply:
modality_lengths is the length of each embedding in X (each one can have different length/dim)

Reply:
Make sure to compute the lengths:
```python
modality_lengths = []
for modality in modalities:
  modality_lengths.append(X.get_column(modality)[0].length()
```